esptool/__init__.py

```diff
@@ -1010,21 +1035,34 @@ def arg_auto_chunk_size(string: str) -> int:
     return num
 
 
-def get_port_list():
+def substr_in_list(portName, filterNames):
+    for s in filterNames:
+        if s in portName:
+            return True
+    return False
+
+
+def get_port_list(vids=[], pids=[], names=[]):
     if list_ports is None:
         raise FatalError(
             "Listing all serial ports is currently not available. "
             "Please try to specify the port when running esptool.py or update "
             "the pyserial package to the latest version"
         )
-    port_list = sorted(ports.device for ports in list_ports.comports())
-    if sys.platform == "darwin":
-        port_list = [
-            port
-            for port in port_list
-            if not port.endswith(("Bluetooth-Incoming-Port", "wlan-debug"))
-        ]
-    return port_list
+    ports = []
+    for port in list_ports.comports():
+        if sys.platform == "darwin" and port.device.endswith(
+            ("Bluetooth-Incoming-Port", "wlan-debug")
+        ):
+            continue
+        if vids and (port.vid is None or port.vid not in vids):
+            continue
+        if pids and (port.pid is None or port.pid not in pids):
+            continue
+        if names and (port.name is None or not substr_in_list(port.name, names)):
```

Comment:
This substr_in_list is very specific and with inverting the result it is pretty hard to understand.  Could we replace it with an inline check similar to the following?

```python
... or any(name not in port.name for name in names)
```

Reply:
And I see I've made a mistake. You corrected it to `not any()`. That could be `all()` instead. You can ignore this if you won't make any more changes to this PR.